useMapsAsMQs fails intermittently

Here's the full exception from the `useMapsAsMQs` test, at commit e3c1c7cb77bafb486fa4a1ff96d69f34f58cd692, which is from PR #778, which is based at `master` at commit b94c64dab34516ef0463310ef656d75b58bad417

    useMapsAsMQs                                                [FAIL - CancellationException:L148]
    [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 120.001 s <<< FAILURE! - in org.corfudb.runtime.concurrent.MapsAsMQsTest
    [ERROR] useMapsAsMQs(org.corfudb.runtime.concurrent.MapsAsMQsTest)  Time elapsed: 120.001 s  <<< ERROR!
    java.util.concurrent.CancellationException
            at java.util.concurrent.FutureTask.report(FutureTask.java:121)
            at java.util.concurrent.FutureTask.get(FutureTask.java:192)
            at org.corfudb.AbstractCorfuTest.executeScheduled(AbstractCorfuTest.java:353)
            at org.corfudb.AbstractCorfuTest.executeScheduled(AbstractCorfuTest.java:315)
            at org.corfudb.runtime.concurrent.MapsAsMQsTest.useMapsAsMQs(MapsAsMQsTest.java:148)
            at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
            at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
            at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.lang.reflect.Method.invoke(Method.java:498)
            at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
            at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
            at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
            at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
            at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
            at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
            at org.corfudb.AbstractCorfuTest$1$1.evaluate(AbstractCorfuTest.java:76)
            at org.junit.rules.RunRules.evaluate(RunRules.java:20)
            at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
            at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
            at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
            at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
            at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
            at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
            at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
            at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
            at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
            at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
            at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
            at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
            at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
            at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
            at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
            at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)

--- a/test/src/test/java/org/corfudb/runtime/concurrent/MapsAsMQsTest.java
+++ b/test/src/test/java/org/corfudb/runtime/concurrent/MapsAsMQsTest.java
@@ -7,7 +7,7 @@ import org.junit.Test;
 
 import java.util.Map;
 import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.locks.Condition;
+import java.util.concurrent.Semaphore;
 import java.util.concurrent.locks.ReentrantLock;
 
 /**
@@ -34,9 +34,8 @@ public class MapsAsMQsTest extends AbstractTransactionsTest {
 
         final int nThreads = 4;
         CountDownLatch barrier = new CountDownLatch(nThreads-1);
-        ReentrantLock lock = new ReentrantLock();
-        Condition c1 = lock.newCondition();
-        Condition c2 = lock.newCondition();
+        Semaphore s1 = new Semaphore(0);
+        Semaphore s2 = new Semaphore(0);
 
 
         // 1st thread: producer of new "trigger" values
@@ -47,23 +46,14 @@ public class MapsAsMQsTest extends AbstractTransactionsTest {
             log.debug("all started");
 
             for (int i = 0; i < numIterations; i++) {
+                // place a value in the map
+                log.debug("- sending 1st trigger " + i);
+                testMap1.put(1L, (long) i);
 
-                try {
-                    lock.lock();
+                // await for the consumer condition to circulate back
+                s2.acquire();
 
-                    // place a value in the map
-                    log.debug("- sending 1st trigger " + i);
-                    testMap1.put(1L, (long) i);
-
-                    // await for the consumer condition to circulate back
-                    c2.await();
-
-                    log.debug("- sending 2nd trigger " + i);
-
-
-                } finally {
-                    //lock.unlock();
-                }
+                log.debug("- s2.await() finished at " + i);
             }
         });
 
@@ -83,14 +73,9 @@ public class MapsAsMQsTest extends AbstractTransactionsTest {
                 log.debug( "- received 1st trigger " + i);
 
                 // 1st producer signal through lock
-                try {
-                    lock.lock();
-
-                    // 1st producer signal
-                    c1.signal();
-                } finally {
-                    lock.unlock();
-                }
+                // 1st producer signal
+                log.debug( "- sending 1st signal " + i);
+                s1.release();
             }
         });
 
@@ -101,21 +86,17 @@ public class MapsAsMQsTest extends AbstractTransactionsTest {
             barrier.countDown();
 
             for (int i = 0; i < numIterations; i++) {
-                try {
-                    TXBegin();
-                    lock.lock();
+                TXBegin();
+                log.debug( "- received 1st condition POST-lock PRE-await" + i);
 
-                    // wait for 1st producer signal
-                    c1.await();
-                    log.debug( "- received 1st condition " + i);
+                // wait for 1st producer signal
+                s1.acquire();
+                log.debug( "- received 1st condition " + i);
 
-                    // produce another tigger value
-                    log.debug( "- sending 2nd trigger " + i);
-                    testMap1.put(2L, (long) i);
-                    TXEnd();
-                } finally {
-                    lock.unlock();
-                }
+                // produce another tigger value
+                log.debug( "- sending 2nd trigger " + i);
+                testMap1.put(2L, (long) i);
+                TXEnd();
             }
         });
 
@@ -133,15 +114,9 @@ public class MapsAsMQsTest extends AbstractTransactionsTest {
                 log.debug( "- received 2nd trigger " + i);
 
                 // 2nd producer signal through lock
-                try {
-                    lock.lock();
-
-                    // 2nd producer signal
-                    log.debug( "- sending 2nd signal " + i);
-                    c2.signal();
-                } finally {
-                    lock.unlock();
-                }
+                // 2nd producer signal
+                log.debug( "- sending 2nd signal " + i);
+                s2.release();
             }
         });
 

